source/common/stats/thread_local_store.h

```diff
@@ -171,13 +172,35 @@ class ThreadLocalStoreImpl : Logger::Loggable<Logger::Id::stats>, public StoreRo
   const Stats::StatsOptions& statsOptions() const override { return stats_options_; }
 
 private:
-  template <class Stat> using StatMap = CharStarHashMap<Stat>;
+  friend class ThreadLocalStoreTestScope;
+
+  template <class Stat> using StatMap = ConstCharStarHashMap<Stat>;
 
   struct TlsCacheEntry {
     StatMap<CounterSharedPtr> counters_;
     StatMap<GaugeSharedPtr> gauges_;
     StatMap<TlsHistogramSharedPtr> histograms_;
     StatMap<ParentHistogramSharedPtr> parent_histograms_;
+
+    // We keep a TLS cache of rejected stat names. This costs memory, but
+    // reduces runtime overhead running the matcher. Moreover, once symbol
+    // tables are integrated, rejection will need the fully elaborated string,
+    // and it we need to take a global symbol-table lock to run. We keep
+    // this char* map here in the TLS cache to avoid taking a lock to compute
+    // rejection.
+    //
+    // To avoid duplicate copies of stats, this map references const char*
+    // values from ThreadLocalStore::rejected_stats_, which is implemented
+    // as a absl::flat_hash_map<char*>, so the keys will be stable and safe
+    // to reference from other threads even as the hash-map resizes.
+    //
+    // Note that once a stat or rejected name enters into the TLS Cache, it
```

Comment:
Can we upgrade this to a TODO? IMO I would consider this a bug for a long running Envoy? WDYT?

Reply:
agreed; done and reported #6306